### ISSUE TITLE
Anamolies with Noto Sans fonts for Indian languages

### ISSUE THREAD

**Issue by tsnbhushan** [wkhtmltopdf/wkhtmltopdf#4949]
**wkhtmltopdf version(s) affected**: 0.12.6 (with patched Qt) 

**OS information**

**Description**

**How to reproduce**

**Expected behavior**

**Possible Solution**

**Comment by PhilterPaper** [comment]
Is the Indian text (in the HTML) just a list of raw Unicode points, rather than glyph IDs? If so, is that what you're seeing on output (just the Unicode, with no processing for glyph reordering, ligatures, etc.)? If so, I suspect that wkHTMLtoPDF's rendering engine may not be using something like HarfBuzz, or may be using a very old version with lots of bugs. Most modern browsers use the HarfBuzz library to properly shape complex scripts such as used in India.

I ran into this problem with the Perl PDF-creation library PDF::Builder, and had to pull in the HarfBuzz::Shaper package to convert complex script raw Unicode characters to glyphs, reorder as necessary, and combine and/or fine tune positioning on glyphs to get decent PDF output. (I don't know any language other than English, so please don't ask me for details!) My guess at this point is that wkHTMLtoPDF's rendering engine needs a major update. There may be a version floating around that uses the current Chrome engine -- maybe give that a try.

**Comment by tsnbhushan** [comment]
Hi,
I am not an expert in the area, so let me try to explain the best I can.  The html contains utf-8 byte-sequences, which may be decoded to their unicode counterparts.  The Noto fonts are designed by Google, and must have all the glyph-related information embedded in them.  The wkhtmltopdf software seems handle most of the reordering/fusion of glyphs correctly, with just a few exceptions, such as those in the example text (ie, the html file in my original ticket).  Someone with familiarity of  Indian scripts should be able to zero in on the specific issues, and assess the impact of resolving those.  

The bottom-line is:  I would like to use a tool that can be run on the command-line (or through a windows batch script), that handles Indian fonts correctly.  Wkhtmltopdf is the closest such tool I have come across.  If the above-mentioned problems can be corrected within wkhtmltopdf, or if there is a command-line (or batch script) method to invoke the same (microsoft) print-to-pdf service that is used by the chrome browser on my machine, that would help me a great deal.

**Comment by PhilterPaper** [comment]
What I'm saying is that the UTF-8 characters are translated into glyphs from the font file. That's as far as you need to go for simple scripts (e.g., Latin alphabet for English), but for complex scripts such as various Indian alphabets, further manipulation is needed: swapping positions of glyphs, consolidating multiple glyphs into a single glyph (ligatures), and tuning the exact positions of glyphs. Browsers typically use the HarfBuzz library to do all this. wkHTMLtoPDF, if it is doing _some_ of this, may simply have a very old version of HarfBuzz (or something similar) in its code. What may be needed is to update wkHTMLtoPDF to use a current rendering engine. As I said, there is supposedly a version of wkHTMLtoPDF based on the current Chrome rendering engine, which may do a better job handling Indic scripts. All you can do is try it out.

**Comment by tsnbhushan** [comment]
Oh I see, thanks for the clarification.  Now, where might I find such an updated version of wkHTMLtoPDF?  I am already running the latest stable version; clearly, trying earlier versions is not likely to help.  Any pointers appreciated.  Thanks

**Comment by PhilterPaper** [comment]
I have not tried the Chrome-based rendering engine myself. It's been mentioned a number of times -- I think it's on GitHub somewhere. You'll have to Google this site for mention of Chrome and wkHTMLtoPDF, in order to find out where it is. I can't guarantee that it's a ready-to-run binary executable, and that you won't have to do some building. Good luck!

**Comment by tsnbhushan** [comment]
Thanks again.  Just for my clarification, are you a part of the wkHTMLtoPDF team, or are you commenting on my issue as a person with general familiarity with HTML to PDF conversion? If it is the latter, can I expect a response from the wkHTMLtoPDF team itself, with more specific guidelines on how to proceed?
BTW, most of the HTML2PDF converters I have tried fall flat when it comes to typesetting Indic fonts.  wkHTMLtoPDF does an almost perfect job, must needs to improve the handling of a few corner-cases. I am hoping that a member of the wkHTMLtoPDF team could help me figure out the best way forward.  Thanks  ...

**Comment by PhilterPaper** [comment]
No, I'm not a developer or a member of the team. I just remember a lot of stuff I've seen go by (I don't know if that's a curse or a blessing). Unfortunately, you very rarely see any responses from the product owner (I don't know if there is anyone else formally on the team). I think you're going to have to move forward yourself, looking around for wkHTMLtoPDF with a Chrome rendering engine, and see if it does a better job for you.

I'll put in a plug for my Perl library PDF::Builder, which does a good job of handling Indic scripts (when combined with HarfBuzz::Shaper). Unfortunately, it's rather low level at this time, unable to handle more than one line at a time (you have to handle the rest of it yourself). Depending on how much control you have over the process, and if you don't _have_ to start with existing HTML, this might be a viable solution for you, so I'll mention it.

**Comment by tsnbhushan** [comment]
Oh I see, thank you for the clarification.  Yes, I would like to try your PDF::Builder as well.  Where can I find it? Is it sufficiently documented?  What is involved in combining it with the (most up-to-date) HarfBuzz?
Thanks ...

**Comment by PhilterPaper** [comment]
You need a Perl installation first (along with Perl programming ability). If you're on Windows, I suggest Strawberry Perl; if on Linux, there's probably a Perl already installed (update if necessary). To get PDF::Builder, go to cpan.org and type in the name, and choose to download the zip file. Or, your Perl may have an install capability (such as "cpan" or "cpanm") that you can tell to grab the package from a CPAN mirror and do the  installation for you. HarfBuzz::Shaper is another Perl package on cpan.org that can be installed separately. PDF::Builder comes with some sample programs to get you started (including HarfBuzz.pl, illustrating the use of HarfBuzz), but most of producing Indic script pages you'll have to figure out. The point is, that PDF::Builder doesn't come "out of the box" ready to typeset your text -- it's a low level library to help you build a PDF. Provide a string of Unicode points, the font to use and its size, etc., and back comes that line on a page.

**Comment by tsnbhushan** [comment]
Thanks again for the detailed information.
I thought you said earlier that PDF::Builder does a good job of handling
Indic fonts.  But if I have to figure out most of the Indic script
production myself, why would I not start with PyPDF instead (which avoids
having to learn Perl,  Harfbuzz etc.)? Does your approach have any
advantage over that?

**Comment by PhilterPaper** [comment]
Well, PDF::Builder (when using HarfBuzz::Shaper) _does_ do a good job of properly handling complex fonts such as Indic scripts. Its limitation is that it can only take one line at a time of Unicode points, turn them into glyphs, rearrange as necessary (into properly arranged Devanagari, etc.), and print them to the PDF file. It does _not_ handle higher level stuff (at this time), such as paragraphs, headings, headers and footers, images, etc. in a smooth manner such as HTML input (although all the lower level pieces _are_ there).

If you wish to start with HTML, to have someone else handle all the nitty gritty of positioning type on the page, wkHTMLtoPDF can (almost) do the job. PyPDF doesn't sound like what you're looking for, but if you're familiar with it, go ahead. My point was that if you don't _have_ to start with HTML, and you don't mind having to get down in the dirt with dividing up your text into lines and positioning them on the page, a library such as PDF::Builder might be suitable.

**Comment by tsnbhushan** [comment]
Oh I see, I understand. I am afraid with this approach, I will be trading
one set of problems for another. The wkhtmltopdf route leaves me just one
step away from my destination, but the other solutions we have discussed
here amounts to taking several steps back, and then trying to find out way
back home.

I thank you for your time, advice and patient explanations, but I was
hoping that when I raise an issue about wkhtmltopdf, someone from that team
would help me find a targeted solution, and also take the opportunity to
rectify a potential flaw in their product. I guess that is too much to
ask...

**Comment by PhilterPaper** [comment]
Assuming that the root of the problem is either an old version of HarfBuzz in the rendering engine, or the use of something homebrewed, that's why I suggested trying to find an offshoot of wkHTMLtoPDF that uses the current Chrome rendering engine. It might be more up-to-date on HarfBuzz usage. Granted, it's not an official release of wkHTMLtoPDF (perhaps a **fork**), but it will have to do. The owner of wkHTMLtoPDF very rarely makes an appearance here, and I don't recall anyone identifying themselves as official development team members. I'm afraid that you may find yourself waiting for a very long time for an official answer, and even longer for a fix to appear. I hope the owner (still @ashkulz,  I think) is working on an official release using Chrome or another updated rendering engine (that would probably fix a _lot_ of problems), but I haven't seen anything official about that.

**Comment by tsnbhushan** [comment]
Thank you again, Phil.  I'll check it out.

Best ...
Bhushan
